types/react-query/index.d.ts

```diff
@@ -1,142 +1,420 @@
+
+export interface BaseQueryOptions {
+    /**
+     * Set this to `true` to disable automatic refetching when the query mounts or changes query keys.
+     * To refetch the query, use the `refetch` method returned from the `useQuery` instance.
+     */
     manual?: boolean;
+    /**
+     * If `false`, failed queries will not retry by default.
+     * If `true`, failed queries will retry infinitely.
+     * If set to an integer number, e.g. 3, failed queries will retry until the failed query count meets that number.
+     */
     retry?: boolean | number;
     retryDelay?: (retryAttempt: number) => number;
     staleTime?: number;
     cacheTime?: number;
     refetchInterval?: false | number;
     refetchIntervalInBackground?: boolean;
     refetchOnWindowFocus?: boolean;
-    onError?: (err: any) => void;
-    onSuccess?: (data: TResult) => void;
+    refetchOnMount?: boolean;
+    onError?: (err: unknown) => void;
     suspense?: boolean;
+}
+
+export interface QueryOptions<TResult> extends BaseQueryOptions {
+    onSuccess?: (data: TResult) => void;
+    onSettled?: (data: TResult | undefined, error: unknown | null) => void;
     initialData?: TResult;
 }
 
-export interface QueryOptionsPaginated<TResult> extends QueryOptions<TResult> {
-    paginated: true;
-    getCanFetchMore: (lastPage: TResult, allPages: TResult[]) => boolean;
+export interface InfiniteQueryOptions<TResult, TMoreVariable> extends QueryOptions<TResult> {
+    getFetchMore: (lastPage: TResult, allPages: TResult[]) => TMoreVariable | false;
 }
 
-export interface QueryResult<TResult, TVariables> {
-    data: null | TResult;
-    error: null | Error;
-    isLoading: boolean;
+export interface QueryResultBase<TResult> {
+    status: 'loading' | 'error' | 'success';
+    error: null | unknown;
     isFetching: boolean;
-    isCached: boolean;
     failureCount: number;
-    refetch: (arg?: {variables?: TVariables, merge?: (...args: any[]) => any, disableThrow?: boolean}) => Promise<void>;
+    refetch: ({ force, throwOnError }?: { force?: boolean; throwOnError?: boolean }) => Promise<TResult>;
 }
 
-export interface QueryResultPaginated<TResult, TVariables> extends QueryResult<TResult[], TVariables> {
-    isFetchingMore: boolean;
-    canFetchMore: boolean;
-    fetchMore: (variables?: TVariables) => Promise<TResult>;
+export interface QueryLoadingResult<TResult> extends QueryResultBase<TResult> {
+    status: 'loading';
+    data: TResult | undefined; // even when error, data can have stale data
+    error: unknown | null; // it still can be error
 }
 
-export function prefetchQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: PrefetchQueryOptions<TResult>
-): Promise<TResult>;
+export interface QueryErrorResult<TResult> extends QueryResultBase<TResult> {
+    status: 'error';
+    data: TResult | undefined; // even when error, data can have stale data
+    error: unknown;
```

Comment:
Isn't it all right to have type Error here? Can it be any other type than that?

Reply:
I think @Igorbek explained it quite well here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43438/files#r399837161

> To clarify, `const { status, error } = useQuery(key, () => throw null)` would produce `status === 'error' && error === null`.

Reply:
yeah but come on, this is very bad, later on in the code is quite useles, to do something like error.message that is what we all probably going to do you need to cast it to Error 
```
<pre>{(error as Error)?.message}</pre>
```
Something like this 😭

Reply:
actually maybe `Error | null` can work fine?

Reply:
I understand the concern and wish thrown exceptions were always of type `Error`, but in JavaScript you can throw and catch anything, including primitives, `null`, `undefined`. This the why when you write `try { ... } catch (err) { ... }` TS doesn't give you the type of the error and just uses `any`. Most of the libraries that propagate exceptions have to decide what to use for error `any` or `unknown`. For example, promises always have `error: any` as a parameter to `onReject`. I chose to use `unknown` here as it was more conservative and puts further decisions on consumers.
`Error | null` wouldn't work either for the same reasons.

Reply:
got it! Sad, but thanks for the explenation

Reply:
I see where you are coming from though. In many other languages async functionality is represented with a result type, for instance: Result<SuccessType, ErrorType = unknown>, that way you can type both success and failure state. It could even default to unknown for the failure state.

Reply:
However, unless this is blocking valid use cases (which I doubt since it simply requires you to cast to a type) then I think this should be added in a patch or minor version later on. I rather have non blocking functional types now that we can extend than 100% correct types in a month.

Reply:
Result ♥️! Anyway no problem at all! Thanks for merging this!